src/passes/Print.cpp

```diff
@@ -119,59 +131,104 @@ std::ostream& operator<<(std::ostream& os, SigName sigName) {
             }
             os << "]";
             continue;
+          } else if (heapType.isStruct()) {
+            auto struct_ = heapType.getStruct();
+            os << "{";
+            auto sep = "";
+            for (auto& field : struct_.fields) {
+              os << sep;
+              sep = "_";
+              if (field.mutable_) {
+                os << "mut:";
+              }
+              printType(field.type);
+            }
+            os << "}";
+            continue;
+          } else if (heapType.isArray()) {
+            os << "[";
+            auto element = heapType.getArray().element;
+            if (element.mutable_) {
+              os << "mut:";
+            }
+            printType(element.type);
+            os << "]";
+            continue;
           }
         }
         os << t;
       }
     }
   };
 
+  auto type = typeName.type;
+  if (type.isBasic()) {
+    os << type;
+    return os;
+  }
   os << '$';
-  printType(sigName.sig.params);
-  os << "_=>_";
-  printType(sigName.sig.results);
+  if (type.isSignature()) {
+    auto sig = type.getSignature();
+    printType(sig.params);
+    os << "_=>_";
+    printType(sig.results);
+  } else if (type.isStruct()) {
+    auto struct_ = type.getStruct();
+    os << "{";
+    auto sep = "";
+    for (auto& field : struct_.fields) {
+      os << sep;
+      sep = "_";
+      if (field.mutable_) {
+        os << "mut:";
+      }
+      printType(field.type);
+    }
+    os << "}";
+  } else if (type.isArray()) {
+    os << "[";
+    auto element = type.getArray().element;
+    if (element.mutable_) {
+      os << "mut:";
+    }
+    printType(element.type);
+    os << "]";
+  } else {
+    os << type;
+  }
   return os;
 }
 
-// Wrapper for printing a type when we try to print the type name as much as
-// possible. For example, for a signature we will print the signature's name,
-// not its contents.
-struct TypeName {
-  Type type;
-  TypeName(Type type) : type(type) {}
-};
-
-struct HeapTypeName {
-  HeapType type;
-  HeapTypeName(HeapType type) : type(type) {}
-};
-
 std::ostream& operator<<(std::ostream& os, TypeName typeName) {
   auto type = typeName.type;
   if (type.isRef() && !type.isBasic()) {
     os << "(ref ";
     if (type.isNullable()) {
       os << "null ";
     }
-    auto heapType = type.getHeapType();
-    if (heapType.isSignature()) {
-      os << SigName(heapType.getSignature());
-    } else {
-      os << heapType;
-    }
+    os << HeapTypeName(type.getHeapType());
     os << ')';
     return os;
   }
   return os << SExprType(typeName.type);
 }
 
-std::ostream& operator<<(std::ostream& os, HeapTypeName typeName) {
+std::ostream& operator<<(std::ostream& os, ResultTypeName typeName) {
```

Comment:
I'm confused about printing "***Type" vs. "***TypeName" here. This routine seems to be the same as [`ResultType`'s `operator<<`](https://github.com/WebAssembly/binaryen/blob/ef8c9f6d06b29fe2c7fe63236d709e3278a55041/src/wasm/wasm-type.cpp#L871-L873), so it doesn't need to be duplicated..? And in my understanding (correct me if I'm wrong), printing `***TypeName` is something like `${i32_f32}`, while printing `***Type` is something like `(struct (field i32) (field f32))`. And this routine seems to be doing the latter..?

Reply:
This routine prints an inline type at the "top level", and names inside it. That is, it will print a tuple like this: `(result $foo $bar)`. The internal components of the tuple have type names, not the expanded type contents.

The difference is that `ResultType` doesn't use a `TypeName` internally for printing the sub-components. This is like a mixture of `ResultType` plus `TypeName`, so I called it `ResultTypeName`... :smile: 

This could all probably be simplified. I can take a look at the whole set of Printing helper classes and stuff later if that's ok. At least some of this complexity is because tuple types do not always have an `()` around them (here they don't, as in the example above, because it's in a result - this is not a problem in a param, oddly...) - maybe that could be simplified too.

Reply:
I don't think we have a `ParamTypeName` in this file?

Reply:
Oh what I was wondering is, if we need `ResultTypeName`, why don't we also need `ParamTypeName`? Maybe I'm missing something.

Reply:
Oh, sorry for misunderstanding you before. It's because the wat text format is weird here. Both the params and the results are a tuple. But the params are written as `(param X) (param Y)` while the results are written as `(result X Y)`. (Maybe the difference is because params have names, so they each need their own element?) We can just loop over params and print them, but we can't do the same for results - we need to write them as a tuple but without the outer `()`, basically (that is, it would be wrong to print `(result (X Y))`, it should be `(result X Y)`). So we wrap the results in a special way to handle that.

There might be a better approach, I'm not sure offhand of the benefits vs alternatives. I think @tlively wrote that code originally and maybe remembers?

Reply:
Oh I see... Yeah maybe we can improve that, but I don't think it needs to block this from landing anyway.